### DISCUSSION TITLE
How to make `actions` run before `guard`?

### DISCUSSION THREAD

**Discussion by MatanYadaev** [statelyai/xstate#4727] [Q&A]
If a state transition configuration has both `actions` and `guard` properties, it first runs `guard` and only then `actions`.

Is there a way to flip the order? So `actions` will run before `guards`? (without using eventless transitions)

Here's an example you can run:

```ts
const machine = createMachine({
  initial: 'Green',
  context: { someData: null } as { someData: string | null },
  states: {
    Green: {
      on: {
        TOGGLE: {
          target: 'Red',
          actions: assign(() => ({ someData: 'hello' })),
          guard: ({context}) => context.someData === 'hello' // Here, context.someData is still `null`
        }
      }
    },
    Red: {
      on: {
        TOGGLE: {
          target: 'Green',
        }
      }
    }
  }
});

const actor = createActor(machine);

actor.start();

actor.send({
  type: 'TOGGLE'
});

console.log(actor.getSnapshot().value); // Green -- It didn't pass to `Red` because of the guard
```

**Comment by davidkpiano** [comment]
The purpose of a `guard` is to determine if the _entire_ transition should be taken or not. The transition includes all actions and the target state to transition to (if any).

The way to model this with state machines is to transition to a "transient state" (a state that has an `always` transition that immediately goes to another state), and to execute the action in that transition. Alternatively, you can model this as a conditionally raised event:

```ts
on: {
  TOGGLE: {
    actions: enqueueActions(({ enqueue }) => {
      enqueue('someAction');
      if (/* condition */) {
        enqueue.raise({ type: 'TOGGLE.CONFIRM' })
      }
    })
  },
  'TOGGLE.CONFIRM': { target: 'red' }
}
```
